e2e storage: public API for testsuites, support CSIInlineVolume type for generic resource

pohly opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Implementing a test suite was impossible outside of the
k8s.io/kubernetes/test/e2e/storage/testsuites package because all
interfaces and structs used by them were private.

**Special notes for your reviewer**:

As part of revamping the API, genericVolumeTestResource also gets
exported because it is useful for other test suites. Because the
TestResource interface became obsolete a while ago and isn't used
anymore, the new name is just testsuites.VolumeResource.

testpatterns.CSIInlineVolume needs special handling in a few places.
It can now be used in a test pattern for a test suite that uses a
VolumeResource instance.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

pohly committed:
e2e storage: public API for testsuites, support CSIInlineVolume type for generic resource

Implementing a test suite was impossible outside of the
k8s.io/kubernetes/test/e2e/storage/testsuites package because all
interfaces and structs used by them were private.

As part of revamping the API, genericVolumeTestResource also gets
exported because it is useful for other test suites. Because the
TestResource interface became obsolete a while ago and isn't used
anymore, the new name is just testsuites.VolumeResource.

testpatterns.CSIInlineVolume needs special handling in a few places.
It can now be used in a test pattern for a test suite that uses a
VolumeResource instance.

pohly commented:
/sig storage

pohly commented:
/retest

davidz627 commented:
I don't have many comments on the code itself but I want us to think very carefully about the implications of exporting all these methods, types, etc. This has bitten us in the butt before where these external interfaces were unable to be changed because of costlyness (https://github.com/kubernetes/kubernetes/pull/77525#issuecomment-490188352). I would prefer to keep as much of the surface unexported as possible. A wholesale change of everything to an exported type doesn't seem like the right solution here - lets figure out some other way to tackle it.

pohly commented:
> I would prefer to keep as much of the surface unexported as possible.

So let's consider how much we need to export while still allowing external test suites:
* The `TestSuite` interface must be exported, which includes its methods.
* To implement that interface, `TestSuiteInfo` must be exported, otherwise an implementation cannot create that struct.
* `VolumeResource` is used in almost every existing test suite. It is obviously pretty useful and was what I needed to support different test patterns in a custom test suite, so I think this should be public together with `CreateVolumeResource`.

That covers the changes in `base.go`. 

The new test patterns are exported just as the existing ones are.

In short, I don't see what names should not be exported.



One way to test that would be to move the base package and individual test suites into different packages.

pohly commented:
One way to test that the base API is suitable would be to separate it from the code using it. For the sake of minimizing changes it would be best to keep "testsuites" as the API and then move individual testsuites into sub-packages (e.g. `test/e2e/storage/testsuites/volumes.go` -> `test/e2e/storage/testsuites/volumes/volumes.go`).

pohly commented:
> One way to test that the base API is suitable would be to separate it from the code using it.

While that's cleaner, it's also some additional churn in the API. Alternatively, we could have a `_test.go` file in some other package which defines a custom test suite using the `testsuites` API. It doesn't need to run, just build.

msau42 commented:
If we're already going to churn the API, why don't we fix things all in one go?

msau42 commented:
/assign

pohly commented:
> If we're already going to churn the API, why don't we fix things all in one go?

I was just brainstorming about other potential changes. So you want the individual test suites in their own packages?

pohly commented:
BTW, right now the API just gets extended. Custom test suites don't need to be changed to continue doing what they were doing before. Moving into packages would be a breaking API change. I think it would be cleaner, I'm just not sure how important it is.

msau42 commented:
Im not sure we really need to move each testsuite to its own package. I think exporting the structs and methods we need is sufficient

pohly commented:
> Im not sure we really need to move each testsuite to its own package.

Yes, neither am I.

> I think exporting the structs and methods we need is sufficient

So the PR is okay as-is, assuming that it meets those goals?

msau42 commented:
I'm asking my teammate to review this PR because we had to do similar work to export various TestSuite interfaces to implement custom test suites.

pohly committed:
e2e storage: add compile test for public TestSuite API

This will catch accidentally adding a new interface function which
isn't exported. For example, an attempt to implement a new unexported
"foobar()" function then leads to:

test/e2e/storage/testsuites/api_test.go:54:5: cannot use &fakeSuite literal (type *fakeSuite) as type testsuites.TestSuite in assignment:
	*fakeSuite does not implement testsuites.TestSuite (missing testsuites.foobar method)
		have foobar()
		want testsuites.foobar()
FAIL	k8s.io/kubernetes/test/e2e/storage/testsuites [build failed]

pohly commented:
I rebased to resolve a merge conflict in `test/e2e/storage/testsuites/volume_expand.go` and added the build test that I mentioned in https://github.com/kubernetes/kubernetes/pull/85540#issuecomment-557696468.

pohly committed:
bazel update

msau42 commented on test/e2e/storage/testpatterns/testpattern.go:
```diff
@@ -162,6 +186,13 @@ var (
 		FeatureTag: "[sig-windows]",
 	}
+	// NtfsEphemeralVolume is TestPattern for "Ephemeral-volume (ntfs)"
+	NtfsEphemeralVolume = TestPattern{
```
csi is not supported on windows yet. This will probably break windows tests?

pohly replied:
I just added it for the sake of completeness. It's not used anywhere yet.

msau42 commented on test/e2e/storage/testsuites/base.go:
```diff
@@ -62,48 +62,42 @@ type opCounts map[string]int64
-	getTestSuiteInfo() TestSuiteInfo
+	// GetTestSuiteInfo returns the TestSuiteInfo for this TestSuite
+	GetTestSuiteInfo() TestSuiteInfo
 	// defineTest defines tests of the testpattern for the driver.
```
comment also needs to be uppercase

pohly replied:
Shall I fix in a follow-up PR?

msau42 commented on test/e2e/storage/testsuites/api_test.go:
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
2019

pohly replied:
Same here - new PR?

But speaking of the copyright year, is there any process in place for ensuring that this gets updated for existing files? Or is the rule "the first PR to touch an existing file must bump to or add the new year"?

msau42 replied:
I believe the year should be when the file was created, not last modified

msau42 replied:
Let's just make the changes now. I already approved so final lgtm will be quick

msau42 commented:
/approve

pohly commented:
I fixed both commits with case fix in the comment and year and force-pushed.

msau42 commented:
/lgtm
/priority important-soon

pohly commented:
/retest

msau42 commented:
/test pull-kubernetes-e2e-gce-storage-slow
